rasterio/_io.pyx

```diff
@@ -668,16 +668,31 @@ cdef class RasterReader(_base.DatasetReader):
                 max(min(window[1][1] or self.width, self.width), 0)))
 
             if overlap != ((0, 0), (0, 0)):
-                data = self.read(indexes, window=overlap)
+                # Prepare a buffer.
+                window_h, window_w = win_shape[-2:]
+                overlap_h = overlap[0][1] - overlap[0][0]
+                overlap_w = overlap[1][1] - overlap[1][0]
+                scaling_h = float(out.shape[-2:][0])/window_h
+                scaling_w = float(out.shape[-2:][1])/window_w
+                buffer_shape = (int(overlap_h*scaling_h), int(overlap_w*scaling_w))
+                data = np.empty(win_shape[:-2] + buffer_shape, dtype)
+                data = self._read(indexes, data, overlap, dtype)
             else:
                 data = None
 
             if data is not None:
                 # Determine where to put the data in the output window.
-                window_h, window_w = win_shape[-2:]
                 data_h, data_w = data.shape[-2:]
-                roff = -window[0][0]
-                coff = -window[1][0]
+                roff = 0
+                coff = 0
+                if window[0][0] < 0:
+                    roff = int(window_h*scaling_h) - data_h
+                if window[0][0] > 0:
```

Comment:
@sgillies This ```if``` and the one on 694 are unnecessary since they do not change the value of ```roff``` and ```coff```